### PR TITLE
feat[next]: Output argument with non-zero domain start

### PR DESCRIPTION
**Pull request by tehrengruber**
```python
field = gtx.as_field(gtx.domain({IDim: (1, 10)}), arr)
field_operator(out=field)
```

This PR also adds a test for non-zero domain start input arguments, which already worked before.

### CODE REVIEW

File: tests/next_tests/integration_tests/feature_tests/ffront_tests/test_program.py [hunk 1]
```diff
@@ -251,3 +251,40 @@ def empty_domain_program(a: cases.IJField, out_field: cases.IJField):
...
+    out = constructors.empty(
+        common.domain({IDim: (1, size - 2)}),
+        allocator=cartesian_case.allocator,
+    )
+    ref = inp.ndarray[1:-2]
+
+    cases.verify(cartesian_case, copy_program, inp, out=out, ref=ref)
+
+
+def test_in_field_arg_with_non_zero_domain_start(cartesian_case, copy_program_def):
```
Comment (tehrengruber) [inline tests/next_tests/integration_tests/feature_tests/ffront_tests/test_program.py:271]: This already worked before, I've just added a test.

File: src/gt4py/next/ffront/past_process_args.py [hunk 2]
```diff
@@ -83,40 +83,46 @@ def _process_args(
             # TODO(tehrengruber): Previously this function was called with the actual arguments
             #  not their type. The check using the shape here is not functional anymore and
             #  should instead be placed in a proper location.
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/gt4py/next/ffront/past_process_args.py:85]: Yes.

File: src/gt4py/next/ffront/past_process_args.py [hunk 3]
```diff
@@ -83,40 +83,46 @@ def _process_args(
...
-            shapes_and_dims = [*_field_constituents_shape_and_dims(args[param_idx], param.type)]
+            ranges_and_dims = [*_field_constituents_range_and_dims(args[param_idx], param.type)]
             # check that all non-scalar like constituents have the same shape and dimension, e.g.
             # for `(scalar, (field1, field2))` the two fields need to have the same shape and
             # dimension
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/gt4py/next/ffront/past_process_args.py:89]: Yes.

File: src/gt4py/next/ffront/past_process_args.py [hunk 4]
```diff
@@ -83,40 +83,46 @@ def _process_args(
...
-                assert hasattr(arg, "shape") and len(arg.shape) == len(dims)
-                yield (arg.shape, dims)
+                assert (
+                    hasattr(arg, "domain")
+                    and isinstance(arg.domain, common.Domain)
+                    and len(arg.domain.dims) == len(dims)
+                )
+                yield (tuple((r.start, r.stop) for r in arg.domain.ranges), dims)
             else:
                 yield from []  # ignore 0-dim fields
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/gt4py/next/ffront/past_process_args.py:127]: You can figure that out with Hannes :D: https://github.com/GridTools/gt4py/pull/1540#discussion_r1701638305

File: src/gt4py/next/ffront/past_process_args.py [hunk 5]
```diff
@@ -83,40 +83,46 @@ def _process_args(
...
         case ts.ScalarType():
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/gt4py/next/ffront/past_process_args.py:128]: As above.

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_python_codegen.py [hunk 6]
```diff
@@ -84,6 +84,11 @@ def builtin_if(*args: Any) -> str:
     return f"{true_val} if {cond} else {false_val}"
 
 
+def builtin_tuple_get(*args: Any) -> str:
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/gt4py/next/program_processors/runners/dace_fieldview/gtir_python_codegen.py:87]: No clue, this is a question for @edopao.

File: src/gt4py/next/ffront/past_process_args.py [hunk 7]
```diff
@@ -83,40 +83,46 @@ def _process_args(
...
             dims = type_info.extract_dims(arg_type)
             if isinstance(arg, ts.TypeSpec):  # TODO
                 yield (tuple(), dims)
             elif dims:
-                assert hasattr(arg, "shape") and len(arg.shape) == len(dims)
-                yield (arg.shape, dims)
+                assert (
+                    hasattr(arg, "domain")
+                    and isinstance(arg.domain, common.Domain)
+                    and len(arg.domain.dims) == len(dims)
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/gt4py/next/ffront/past_process_args.py:123]: I don't really like coupling the assert with the usage of the domain below. As an assert alone this is nicer indeed.

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_python_codegen.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (edopao) [reply src/gt4py/next/program_processors/runners/dace_fieldview/gtir_python_codegen.py:87]: Right, I could use just the arguments if I change another definition:
```
-def builtin_if(*args: Any) -> str:
-    cond, true_val, false_val = args
+def builtin_if(cond: str, true_val: str, false_val: str) -> str:
     return f"{true_val} if {cond} else {false_val}"
 
 
@@ -93,7 +92,7 @@ def make_const_list(arg: str) -> str:
     return arg
 
 
-GENERAL_BUILTIN_MAPPING: dict[str, Callable[[Any], str]] = {
+GENERAL_BUILTIN_MAPPING: dict[str, Callable[..., str]] = {
     "cast_": builtin_cast,
     "if_": builtin_if,
     "make_const_list": make_const_list,
```

[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/program_processors/runners/dace_fieldview/gtir_python_codegen.py:87]: @edopao Your suggestion looks better to me than the current code, but I don't know if the change does really need to go into this PR or it can be added to another PR to get this one merged ASAP...

**Review by egparedes — APPROVED** [review]
LGTM.

**Review by havogt — DISMISSED** [review]
Just blocking, because I would like to review the changes.

**Review by edopao — APPROVED** [review]
